### PR TITLE
Add an API check/test for docker build -f Dockerfile

### PR DESCRIPTION
**Pull request by duglin**
I noticed that while we have tests to make sure that people don't
specify a Dockerfile (via -f) that's outside of the build context
when using the docker cli, we don't check on the server side to make
sure that API users have the same check done. This would be a security
risk.

While in there I had to add a new util func for the tests to allow us to
send content to the server that isn't json encoded - in this case a tarball.

I also tweaked an error message that wasn't really accurate.

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

**Comment by duglin** [comment]
Forgot to mention that in addition to the new test I also added a check in the code to block people from leaving the build context.

**Comment by LK4D4** [comment]
@duglin Already needs rebase :)

**Comment by duglin** [comment]
that was quick!  Rebase is done.

**Comment by duglin** [comment]
After reading #10323 I switched this to use filepath.

File: integration-cli/docker_utils.go [hunk 1]
```diff
@@ -296,17 +305,16 @@ func sockRequest(method, endpoint string, data interface{}) ([]byte, error) {
...
-
-	req, err := http.NewRequest(method, endpoint, jsonData)
-	req.Header.Set("Content-Type", "application/json")
+	req, err := http.NewRequest(method, endpoint, data)
 	if err != nil {
 		return nil, fmt.Errorf("could not create new request: %v", err)
 	}
 
+	if ct == "" {
+		ct = "appliation/json"
```
Comment (estesp) [inline integration-cli/docker_utils.go]: minor typo: "`application/json`"

Comment (duglin) [inline integration-cli/docker_utils.go]: fixed - thanks!

**Comment by duglin** [comment]
@jfrazelle ping

**Comment by jfrazelle** [comment]
woohooo tests I love tests LGTM

**Comment by thaJeztah** [comment]
Should this be in the 1.5 milestone, given that this (`-f`) is a new feature in Docker 1.5?

**Comment by duglin** [comment]
@thaJeztah good point. W/o the check it in 1.5 its a bit of security risk. @icecrime any chance of adding it?

**Comment by icecrime** [comment]
Agree, it looks necessary for 1.5 :+1:

**Comment by icecrime** [comment]
LGTM (@LK4D4 @tiborvass: are you available for a 3rd look?)

File: builder/evaluator.go [hunk 2]
```diff
@@ -169,12 +169,19 @@ func (b *Builder) Run(context io.Reader) (string, error) {
 
 // Reads a Dockerfile from the current context. It assumes that the
 // 'filename' is a relative path from the root of the context
-func (b *Builder) readDockerfile(filename string) error {
-	filename = path.Join(b.contextPath, filename)
+func (b *Builder) readDockerfile(origFile string) error {
+	filename := filepath.Join(b.contextPath, origFile)
+
+	tmpDockerPath := filepath.Dir(filename) + string(os.PathSeparator)
+	tmpContextPath := filepath.Clean(b.contextPath) + string(os.PathSeparator)
```
Comment (crosbymichael) [inline builder/evaluator.go]: why not `filepath.Join`????

Comment (duglin) [inline builder/evaluator.go]: I'm not joining a path and a name - I'm just making sure that these two paths end with /'s

Comment (tiborvass) [inline builder/evaluator.go]: @duglin why do you need Clean here? Is there a specific case you had in mind?

File: integration-cli/docker_api_containers_test.go [hunk 3]
```diff
@@ -299,3 +299,35 @@ func TestGetContainerStats(t *testing.T) {
...
+		t.Fatalf("failed to write tar file header: %v", err)
+	}
+	if _, err := tw.Write([]byte("FROM ubuntu")); err != nil {
+		t.Fatalf("failed to write tar file content: %v", err)
+	}
+	if err := tw.Close(); err != nil {
+		t.Fatalf("failed to close tar archive: %v", err)
+	}
+
+	out, err := sockRequestRaw("POST", "/build?dockerfile=../Dockerfile", buffer, "application/x-tar")
```
Comment (erikh) [inline integration-cli/docker_api_containers_test.go]: Test the absolute path case as well, please.

**Comment by icecrime** [comment]
*Clearing milestone so we can focus on the replacement PR #10513*

**Comment by tiborvass** [comment]
Closing in favor of #10524

**Pull request closed without merging**